### PR TITLE
Fix toggle optimizer

### LINKED ISSUE #5292: For multiple optimizers, model.toggle_optimizer() is setting the requires_grad=True for all the params.
https://github.com/PyTorchLightning/pytorch-lightning/blob/dabfeca92e0702e55f09ac53e9412672cd258cd3/pytorch_lightning/core/lightning.py#L1152-L1170

Setup : 
pytorch-lightning 1.1.2
pytorch 1.7.1

When using multiple optimizers, in the `toggle_optimizers(..)` function, the requires_grad property is set to true for all the params belonging to the param_groups of the optimizer. This is incorrect as in case the user has explicitly disabled the `requires_grad` property for some parameters permanently, the function would enable `requires_grad` for those parameters also. 

Proposed fix:
The requires_grad value for all the parameters should be stored beforehand in an object variable `self.params_dict` of lightning object. In the `toggle_optimizer()` function, instead of setting `params.requires_grad=True`, set `params.requires_grad = self.params_dict[params]`.This would set the correct value for the params of the concerned optimizer.

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #5292 
This more closely follows the suggestions from the issue. Inside of `toggle_optimizers`, first we scan over all `requires_grad` values for all the parameters and store them on the lightning module., and set `requires_grad=False`. Then we iterate over the values of the current optimizer's parameters and set them according to the dictionary created from the initial scan.

The issue with the initial fix was multiple optimizers with shared parameters were handled incorrectly. 

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
         """
+
+        # Iterate over all optimizer parameters to preserve their `requires_grad` information
+        # in case these are pre-defined during `configure_optimizers`
         param_requires_grad_state = {}
-        # make sure current optimizer is latest to be iterated over.
-        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
-        num_optimizers = len(optimizers) - 1
-        for opt_idx, opt in enumerate(optimizers):
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1183]: Can `param_requires_grad_state` be calculated once per training step instead of once per optimizer per training step?

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
             for group in opt.param_groups:
                 for param in group['params']:
-                    if num_optimizers == opt_idx:
-                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
-                        if param in param_requires_grad_state:
-                            param.requires_grad = param_requires_grad_state[param]
+                    # If a param appears in multiple optimizers, use the `requires_grad` info already set
+                    if param in param_requires_grad_state:
+                        param.requires_grad = param_requires_grad_state[param]
```
Comment (s-rog) [inline pytorch_lightning/core/lightning.py:1188]: This forces all optimizers to have the same grad state for all params?

**Comment by Borda** [comment]
@ananthsub could you please add a test so we won't accidentally break it again...

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1183]: It is possible, but it will make use of `toggle_optimizer` slightly harder to properly use for new users.

Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1183]: Shouldn't you exclude the current `optimizer` from this iteration ?

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
-        for opt_idx, opt in enumerate(optimizers):
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
             for group in opt.param_groups:
                 for param in group['params']:
-                    if num_optimizers == opt_idx:
-                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
-                        if param in param_requires_grad_state:
-                            param.requires_grad = param_requires_grad_state[param]
+                    # If a param appears in multiple optimizers, use the `requires_grad` info already set
+                    if param in param_requires_grad_state:
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1187]: Could we have a test with 3 optimizers and shared parameters ? I think the previous code made a 2 optimizers assumption, but this new code works for more.

The assumption was done there, where 3 optimizers could share the same parameters, and by setting  `param.requires_grad = False` with first optimizer, the second one will override the state and we lose the information.
```
param_requires_grad_state[param] = param.requires_grad		                         
param.requires_grad = False
```

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:1188]: Not sure about this line, shouldn't just to be `continue` ? 
With 3 optimizers,
If param is shared only with 1 and 2 optimizers, and current is 3, it is possible you will reset param to True when it should have been False.
Can you check this ?

**Review by tchaton — COMMENTED** [review]
Hey @ananthsub, 

I apologise this part wasn't better tested.
Would you mind to add more tests as it is a pretty critical part of the code.

Best,
T.C

File: tests/core/test_lightning_module.py [hunk 4]
```diff
@@ -211,3 +212,144 @@ def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, clos
...
+                assert self.layer_3[3].weight.requires_grad is True
+                assert self.layer_3[5].weight.requires_grad is False
+
+            optimizer.step(closure=closure)
+
+        def training_step(self, batch, batch_idx, optimizer_idx=None):
+            return super().training_step(batch, batch_idx)
+
+        @staticmethod
+        def combine_genetators(gen_1, gen_2):
```
Comment (SeanNaren) [inline tests/core/test_lightning_module.py:315]: Small spelling mistake in the function name

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
             for group in opt.param_groups:
                 for param in group['params']:
-                    if num_optimizers == opt_idx:
-                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
-                        if param in param_requires_grad_state:
-                            param.requires_grad = param_requires_grad_state[param]
+                    # If a param already appear in param_requires_grad_state, continue
+                    if param in param_requires_grad_state:
+                        continue
                     else:
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1189]:
```suggestion
```

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
-                        # save requires_grad for later restoration
                         param_requires_grad_state[param] = param.requires_grad
                         param.requires_grad = False
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1191]:
```suggestion
                    param_requires_grad_state[param] = param.requires_grad
                    param.requires_grad = False
```

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1176,22 +1176,25 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
             optimizer: Current optimizer used in training_loop
             optimizer_idx: Current optimizer idx in training_loop
         """
+
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1179]: hey can we use the `opt_idx` to get the optimizer down there??... just to unify the arguments with `untoggle_optimizer`

File: tests/core/test_lightning_module.py [hunk 8]
```diff
@@ -211,3 +212,144 @@ def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, clos
     )
 
     trainer.fit(model)
```
Comment (rohitgr7) [inline tests/core/test_lightning_module.py:214]:
```suggestion
        limit_val_batches=0,
    )

    results = trainer.fit(model)
    assert results
```
just limiting the validation.

File: tests/core/test_lightning_module.py [hunk 9]
```diff
@@ -211,3 +212,144 @@ def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, clos
...
+            return [optimizer_1, optimizer_2, optimizer_3]
+
+    model = TestModel()
+    model.training_epoch_end = None
+
+    trainer = Trainer(
+        max_epochs=1,
+        default_root_dir=tmpdir,
+        limit_train_batches=8,
+        accumulate_grad_batches=1,
```
Comment (rohitgr7) [inline tests/core/test_lightning_module.py:352]:
```suggestion
        default_root_dir=tmpdir,
        max_steps=8,
```

File: tests/core/test_lightning_module.py [hunk 10]
```diff
@@ -211,3 +212,144 @@ def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, clos
...
+    )
+
+    trainer.fit(model)
```
Comment (rohitgr7) [inline tests/core/test_lightning_module.py:355]:
```suggestion
    results = trainer.fit(model)
    assert results
```

**Review by rohitgr7 — APPROVED** [review]
nice!

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1179]: n00b q: why do we need it when the current optimizer is already passed in?

**Comment by ananthsub** [comment]
thanks @tchaton for the test cases - confirming that this patch works for me

File: pytorch_lightning/core/lightning.py (hunk 7 above)
Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py:1179]: Nah we don't. either `optimizer` or `optimizer_idx` is enough. `optimizer_idx` seems good for both.

**Review by tchaton — APPROVED** [review]
